# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: widget/gtk/GtkCompositorWidget.cpp
# Commit: 7ac989d08efb
# Full Hash: 7ac989d08efb12bad409cc29984c58c56a04a6b1
# Author: stransky <stransky@redhat.com>
# Date: 2021-01-22 17:29:50
# Regressor Bug: 1674473
# File Overlap Count: 1
# Description:
#   Bug 1674473 [Linux] SW-WR Clear background when painting popup windows, r=lsalzman
#   
#   Clear alpha from buffer before painting transparent windows by SW Webrender to avoid rendering artifacts.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D102189
# ==============================================================================

diff -r 3046f61ecff2 -r 7ac989d08efb widget/gtk/GtkCompositorWidget.cpp
--- a/widget/gtk/GtkCompositorWidget.cpp	Fri Jan 22 08:10:12 2021 +0000
+++ b/widget/gtk/GtkCompositorWidget.cpp	Wed Jan 20 21:11:48 2021 +0000
@@ -14,6 +14,8 @@
 namespace mozilla {
 namespace widget {
 
+using namespace gfx;
+
 GtkCompositorWidget::GtkCompositorWidget(
     const GtkCompositorWidgetInitData& aInitData,
     const layers::CompositorOptions& aOptions, nsWindow* aWindow)
@@ -125,5 +127,20 @@
 }
 #endif
 
+void GtkCompositorWidget::ClearBeforePaint(
+    RefPtr<gfx::DrawTarget> aTarget, const LayoutDeviceIntRegion& aRegion) {
+  // We need to clear target buffer alpha values of popup windows as
+  // SW-WR paints with alpha blending (see Bug 1674473).
+  if (mWidget->IsPopup()) {
+    for (auto iter = aRegion.RectIter(); !iter.Done(); iter.Next()) {
+      LayoutDeviceIntRect r = iter.Get();
+      aTarget->FillRect(gfx::Rect(r.x, r.y, r.width, r.height),
+                        ColorPattern(DeviceColor(0, 0, 0, 0)),
+                        DrawOptions(1.0f, CompositionOp::OP_SOURCE));
+    }
+    aTarget->Flush();
+  }
+}
+
 }  // namespace widget
 }  // namespace mozilla